### ISSUE TITLE
Do binary search in findLeafNodeAtOffset

### ISSUE THREAD

**Issue by spoenemann** [langium/langium#630]
The CST is a sorted tree. It should be possible to implement `findLeafNodeAtOffset` (`cst-util.ts`) in logarithmic time with a binary search algorithm.

**Comment by snarkipus** [comment]
I'll take a shot at this unless there's an objection.

**Comment by snarkipus** [comment]
Tags: @msujew @spoenemann 
## Code Review Request
Here's a cut ... I'm not terribly fond of it, but I do think it's binary'ish. It passes all the tests, but I didn't go to the extent of actually trying to profile it. It just feels like I'm missing something that would make it a bit more clean, but maybe not.

Thoughts?

```TypeScript
export function findLeafNodeAtOffset(node: CstNode, offset: number): LeafCstNode | undefined {
    if (isLeafCstNode(node)) {
        return node;
    } else if (isCompositeCstNode(node)) {
        let firstChild = 0;
        let lastChild = node.children.length - 1;
        let middleChild = Math.floor((firstChild + lastChild)/2);

        do {
            if (node.children[middleChild].offset === offset) {
                return unfoldChild(node.children[middleChild], offset);
            }
            if (node.children[middleChild].offset > offset) {
                lastChild = middleChild - 1;
            } else {
                firstChild = middleChild + 1;
            }
            middleChild = Math.floor((firstChild + lastChild)/2);
        } while (firstChild <= lastChild);
        return unfoldChild(node.children[middleChild], offset);
    }
    return undefined;
}

function unfoldChild(node: CstNode, offset: number): LeafCstNode | undefined {
    switch (isLeafCstNode(node)) {
        case true: return node as LeafCstNode;
        case false: return findLeafNodeAtOffset(node, offset);
        default: return undefined;
    }
}
```

**Comment by spoenemann** [comment]
Thanks @snarkipus! Thinking about ways to simplify the code, I found it the easiest to write a suggestion for the `isCompositeCstNode` case based on your draft:

```typescript
let firstChild = 0;
let lastChild = node.children.length - 1;

while (firstChild <= lastChild) {
    const middleChild = Math.floor((firstChild + lastChild) / 2);
    const n = node.children[middleChild];
    if (n.offset > offset) {
        lastChild = middleChild - 1;
    } else if (n.end <= offset) {
        firstChild = middleChild + 1;
    } else {
        return findLeafNodeAtOffset(n, offset);
    }
}
```

This is totally untested, feel free to check it out and submit a PR.

**Comment by spoenemann** [comment]
Fixed with #657
